server: remove possible causes of deadlocks

stevendanna opened the pull request:
See the individual commits for details.

Epic: CRDB-14537
Release note: None

Co-authored-by: Raphael 'kena' Poss <knz@thaumogen.net>

stevendanna committed:
server: quiesce tenant stopper when server stops

Previously, the tenant stopper would only be stopped by a call to
serverController.Stop(). However, that Stop() call would only happen
if the top level stopper had quiesced, which it couldn't do because it
was waiting on the tenant server to start up.

The problem was caused by a mistaken assumption that the closers
on the top level stopper are run before  Quiesce(),
whereas it's the other way around: Quiesce() waits for all tasks
to terminate before calling the closers.

This patch fixes this situation by hoisting the instantiation of the
tenant stopper to the control function, and ensuring that the
top-level stopper quiescence propagates to a Stop() call on the child
stopper.

Additionally, we discovered that the KV tenant connector was not
properly aborting its own startup upon server shutdown. This was an
oversight from the original implementation. This patch fixes that by
adding the proper check on the stopper ShouldQuiesce() channel.

Release note: None

stevendanna committed:
pkg/server: close cmux listener and don't retry accept errors

This patch fixes an ancient bug, which is somewhat hidden
(we don't see it triggered in production easily): the cmux listener
was not being closed properly upon server shutdown.

This tends to be invisible in practice because in a real server the
main goroutine (in cli/start) terminates and brings the process away
with it without waiting for this. However, for the sake of generality
we want to tie up the loose ends. It also reduces uncertainty about
possible cmux-related deadlocks when shared-process tenant servers are
shut down early.

Release note: None

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r1, 1 of 1 files at r2, 1 of 2 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96268#-:-NN728305Ijzp0wZFd_Z:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @cucaroach and @msbutler)



<!-- Sent from Reviewable.io -->

stevendanna committed:
server: avoid deadlock during fast shutdown

When dealing with on-demand started tenants in tests, we may see a
server shutdown during a tenant start up.

During server start up, this code adds a closer which calls
grpc.Stop() and also, as a work-around to another bug, calls
(*cmux).Serve().  It assumed that by the time Serve() was called, the
listeners would all be closed and thus Serve would fail with an error.

This assumption was wrong. The listeners were only closed in an async
task that was started _after_ the closer was installed. Typically this
is fine, but in the case of a stopper that is already stopping, the
AddCloser call runs the given closer immediately. As a result, the
closer was waiting forever in cmux.Serve.

Now, we only install the closer after the asyncronous task is started.
Further, if the async task fails to start, we close the listeners
directly.

Release note: None

Epic: none

knz reviewed (commented):
> Reviewed 1 of 2 files at r5, 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96268)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @msbutler)



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> Reviewed 1 of 2 files at r8, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/96268)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @cucaroach and @msbutler)



<!-- Sent from Reviewable.io -->

knz commented:
bors r+ p=999 single on
